### ISSUE TITLE
Debugger: Add support for "pause on unhandled rejection"

### ISSUE THREAD

**Issue by chrisdickinson** [nodejs/promises#15]
By request from @trevnorris, it would be nice to pause the debugger on unhandled promise rejection.

**Comment by benjamingr** [comment]
cc @petkaantonov what would you need to take a look into this?

@paulirish any chance you can hook us up with the relevant info or the people that have the relevant info?

**Comment by petkaantonov** [comment]
By the time unhandled rejection is emitted the original stack is no longer the current stack. So this would also required the ability to save stack traces so that they can be inspected after the fact as if they were the current stack.

**Comment by chrisdickinson** [comment]
We can wire this up using [`SetPromiseRejectCallback`](https://github.com/v8/v8/blob/master/include/v8.h#L5827-L5831) without any new APIs, I believe.

**Comment by trevnorris** [comment]
@petkaantonov stack traces don't come near to providing enough information in a complex scenario. Hence's a contrived, but simple, example of why knowing immediately when something threw is important:
```js
new Promise(function(res, rej) {
  const arr = [];
  while (true)
    arr.push(new Buffer(0xfffffff).fill(0));  // alloc until OOM
}).catch(function(er) {
  console.log(er.message);
  gc();
  console.log(process.memoryUsage());
});

// Output:
// Array buffer allocation failed
// { rss: 24264704, heapTotal: 8433664, heapUsed: 3443728 }
```
No one I've seen takes into consideration that `new Buffer()` may throw from OOM. Knowing exactly when this happens is very useful to analyse the exact state of the application. Whether it has actually exhausted application resources, or if an external application tied up memory. But waiting until later means we'll have lost state possibly critical to properly debug the error.

In the case a user has explicitly placed a try/catch around `new Buffer` I'd assume they know what they're looking for. But as I mentioned, node developers don't write code in anticipation that OOM may happen. In cases like this I consider it critical to peel back and be allowed to stop execution on the thrown exception. Even if it's within a Promise.

**Comment by petkaantonov** [comment]
I didn't realize native OOM even throw a catchable error. When you run out of js memory, the process will just abort directly from V8 and no catchable exception is thrown. Why is the behavior for native oom different from js oom?

**Comment by benjamingr** [comment]
> No one I've seen takes into consideration that new Buffer() may throw from OOM. 

This is a great example of a synchronous operational error by the way. (cc @chrisdickinson )

**Comment by trevnorris** [comment]
@benjamingr Can you expand on what you're getting at?

**Comment by benjamingr** [comment]
@trevnorris oh sorry - that was out of context. We were discussing operational vs. programmer errors earlier. I was looking for more examples (like `JSON.parse` or `.listen`) that are operational errors that happen synchronously. It's unrelated to this discussion - I just thought it was a very good example so I pinged Chris. Sorry for the offtopic :)

**Comment by trevnorris** [comment]
TL;DR Handling an exception that occurs from OOM at any point besides the moment it was thrown compromises state and makes debugging less, if at all, useful.

@petkaantonov I'll also make this conjecture about why it handles memory differently: When v8 aborts from running out of heap space it will have aggressively GC'd and done anything else it can before giving up the ghost. Whereas external user allocated memory hasn't had that chance. So v8 simply throws an error that the operation couldn't complete. In the case the dev can do manual resource cleanup and attempt the allocation again.

This has actually happened to me. Was reading in a very large number of files into memory to perform some analysis on them. It was considerably faster to store them all in memory while randomly accessing the data. Though this would push my machine to the limit before reaching swap.

This example of what I did is a poor balance between simplicity and only adding the "essentials". Here are the key things to look for:
1) I was lazy and am performing plenty of `*Sync` operations to simply the code.
2) Memory is allocated in 2 locations to:
2.1) Initially over allocate to make sure everything can be read (since this is actually a race condition we have to guard against).
2.2) If over allocated too much then create smaller Buffer and move contents over. Allowing GC to collect the larger object.
3) The unimplemented `cleanupMemory()` function that may free enough memory for our application to continue. For example if we are lazily hanging onto old records, they may be free'd and GC'd.
4) The recovery paths (minus the lazy sync calls) to make sure application state is always happy and recoverable.
```js
'use strict';

const fs = require('fs');
const SlowBuffer = require('buffer').SlowBuffer;
const files_map = new Map();

function processPaths(path_list, cb) {
  // Will pop() array elements off when properly processed. So just
  // need to loop until all elements are gone.
  while (path_list.length > 0) {
    const path = path_list[path_list.length - 1];
    const stats = fs.statSync(path);
    const fd = fs.openSync(path, 'r');
    let buf;

    try {
      // The kernel can lie about file sizes, so grab some extra.
      buf = new Buffer(stats.size + stats.blksize);

    } catch (e) {
      // Assume it's OOM. Attempt cleanup and retry.
      fs.closeSync(fd);
      cleanupMemory();
      return setImmediate(processPaths, path_list, cb);
    }

    const bytesRead = fs.readSync(fd, buf, 0, buf.length, 0);
    if (buf.length === bytesRead) {
      // Possibly didn't read all the data, not covering this.
    }

    // If > 512 bytes are remaining, move data to new Buffer to
    // conserve memory.
    if (buf.length - bytesRead > 512) {
      let sbuf;
      try {
        sbuf = new SlowBuffer(bytesRead);
      } catch (e) {
        // Assume it's OOM. Attempt cleanup and retry.
        fs.closeSync(fd);
        cleanupMemory();
        return setImmediate(processPaths, path_list, cb);
      }
      buf.copy(sbuf, 0, 0, bytesRead);
      buf = sbuf;
    }

    // Everything's completed. Store Buffer to Map.
    files_map.set(path_list.pop(), buf);
    fs.closeSync(fd);
  }
  cb();
}

function cleanupMemory() { /* not implemented */ }
```
Thanks for getting through that with me. Will be happy to clarify any points.

First to address is the allocation of Buffers. This happens so little, unless on a Pi, that no one has remembered to document that Buffer can throw in the case of OOM. Though these cases are very important to hit immediately. Here are a couple reasons for OOM:

1) System has been greedy and not left enough for `new Buffer()`. Not many easy steps we can follow to recover from this.
2) Execution scope went all strange, which filled memory with Buffer instances. Problem is if we aren't alerted to the error until some point in the future then all the allocations from that execution scope will most likely have been GC'd. Thus leaving us without critical information for debugging.
3) Resources are spread around the application taking up memory. Between the exception and reporting internal state could have changed enough by moving, free'ing, etc. that again we loose important information for debugging.

The only viable option for debugging this case properly is to be able to break the moment the exception is thrown.

Phew, done.

**Comment by petkaantonov** [comment]
I think this is really breaking the expectations people have with garbage collected langugage. Having to try catch a OOM error and then perform GC() manually is just simply wrong. This must be fixed in core instead of pushing it to unsuspecting users. What if we used the new cleanup stuff in the workers PR to cleanup stuff so that we can either allocate the buffer or abort the process?

**Comment by trevnorris** [comment]
> Having to try catch a OOM error and then perform GC() manually is just simply wrong.

This isn't the only path. Another real world example is having multiple processes running, and starting a snapshot on one of those. Some snapshots can take up multiple GB of memory. Should we crash the node process simply because it'll be OOM for a little while? Users need to be aware of their entire  machine when it comes to this. The problem may have nothing at all to do with the node instance.

> This must be fixed in core instead of pushing it to unsuspecting users.

I'm curious as to what you believe core could fix in these cases? We can force GC ourselves, but that doesn't imply there are user held resources that need to be manually closed. Thinking that we have to make this fluffy friendly is a misconception. Hitting this scenarios hardly ever happens, unless you're on a Pi. Even so there's not enough core can do by inference of the state of the application that can be done by the dev themselves.

> What if we used the new cleanup stuff in the workers PR to cleanup stuff so that we can either allocate the buffer or abort the process?

Could help, but also see above. Aborting also doesn't seem to be a viable option. For example, here's `d8`:
```
d8> new ArrayBuffer(1024 * 1024 * 1024 * 16)
(d8):1: RangeError: Array buffer allocation failed
new ArrayBuffer(1024 * 1024 * 1024 * 16)
^
RangeError: Array buffer allocation failed
    at (d8):1:1
```
Tried to cause the same failure on Chrome, but that happily kept going right into my 16 GB of swap. Basically bringing down my machine.

**Comment by petkaantonov** [comment]
The Chrome error is different because you can never even allocate that amount of memory while in node in your example there is enough memory it's just that the user must manually call GC() first, which is something that node can easily do automatically. Chrome is not bringing down your machine but throwing the error according to spec: https://tc39.github.io/ecma262/#sec-createbytedatablock

**Comment by Fishrock123** [comment]
> I think this is really breaking the expectations people have with garbage collected language.

Is it though? I would like to see an example from other languages. You can't go willy-nilly allocating memory expecting nothing to happen. (Unless you are a browser, but we aren't a browser, and eating up swap indefinitely is _horrible_ UX.)

> Having to try catch a OOM error and then perform GC() manually is just simply wrong.

GC won't even necessarily help in this situation if I understand correctly? You could still be anting to hold these even when the system if out of memory for external reasons.

**Comment by petkaantonov** [comment]
The browser isn't eating swap, I literally just tried on this page and it crashed the tab from allocating 2GB even though I have 16GB  ram alone.

**Comment by petkaantonov** [comment]
> You can't go willy-nilly allocating memory expecting nothing to happen.

I expect memory I am no longer using to be freed without me manually having to call GC to do it, how is this even a debate?

**Comment by Fishrock123** [comment]
Sorry, I probably should have clarified:

If you are allocating memory AND holding references to it.

**Comment by trevnorris** [comment]
> The browser isn't eating swap, I literally just tried on this page and it crashed the tab from allocating 2GB even though I have 16GB ram alone.

My bad. Wrong code:
```js
var arr = []; while (true) arr.push((new Uint8Array((-1 >>>0) >>> 1)).fill(1));
```
Have fun swapping your machine to death.

>  Chrome is not bringing down your machine but throwing the error according to spec: https://tc39.github.io/ecma262/#sec-createbytedatablock

Then I'm confused what our argument is. We are throwing. The spec says to throw. Seems like there's no issue there.

**Comment by petkaantonov** [comment]
The difference is that when we throw, we didn't even try to see if we could free up memory for the buffer.

**Comment by Fishrock123** [comment]
I think this is now off-topic, can we make a new thread?

**Comment by petkaantonov** [comment]
I just get TypeError invalid array length when I try to run your example

**Comment by petkaantonov** [comment]
I don't mind making a new thread

**Comment by benjamingr** [comment]
> really long post with OOM example
@trevnorris 

That's a very good post, and it does show us something interesting - so thanks for that :)

I agree that an OOM error is in fact not realistically recoverable from in most cases. I also agree that it's completely important to catch it and get a handle on the memory when it happened and not later.

It is however also an operational error - the only reason it `throw`s is because getting memory in Node is a synchronous operation. The distinction I've made above is because people from the post mortem group are distinguishing programmer errors (crash) from operational errors (recover) by the code `throw`ing or `(err, null)` ing the callback (do correct me if I'm wrong here).

I think we should have a flag for `--abort-on-oom` or something similar in case there already is one. It's an operational error but it's important to get a memory snapshot from when it happened exactly because of the type of error it is. Opinions?

**Comment by chrisdickinson** [comment]
@benjamingr We should move that to a topic on the main issue tracker, since it's not strictly related to promises.

It seems folks have been talking about Chrome Dev Tools' ability to step back into async frames in this discussion — which is interesting and should be pursued in another issue. This issue tracks the work of making the debugger pause on unhandled rejection, which should be possible to implement with the existing APIs we have at our disposal.

**Comment by trevnorris** [comment]
Apologies for allowing the example to control the direction of the conversation. The point I'm making is that there are cases where it's important to know immediately when an error is thrown, but also that the error not simply abort the process. Though to not again accidentally change the topic I'll not post any more examples.

@petkaantonov You must have run that code with a v4.x release of node? Because the error message has changed to "Array buffer allocation failed" in v5.x. Also, did you try running that in Chrome's dev tool bar?

**Comment by petkaantonov** [comment]
I am running everything in Chrome 48 devtools on this page
